### PR TITLE
Refactored loading, plus 

### LINKED ISSUE #12: Game crashes on turn end when deck is empty
Possibly related to #11

### PR DESCRIPTION
**Pull request by keltoff**
- Loading .json separated from rest of the code
- Globals changed to eponymous enum
-- also they have now defined order in UI
- Board zones identified by enum, not String
- Various places with cards changed to struct CardContainer
- Modified widget code to work with CardContainer and be more generic

**Warning**, possibly breaks Globals rendering ( #9 revived plus globals buttons overlap cards)

### CODE REVIEW

**Comment by keltoff** [comment]
Incidentally, on_turn_begin, on_turn_end, and on_strike effects (on buildings and kaiju respectively) might now work. Was not tested very much.

File: src/game_objects.rs [hunk 1]
```diff
@@ -44,80 +87,85 @@ pub struct Card {
...
+
+    pub fn add(&mut self, card: Card) {
+        self.cards.push(card)
+    }
+
+    pub fn empty(&self) -> bool {
+        self.cards.is_empty()
+    }
+
+    pub fn break_one(&mut self) {
```
Comment (Lycowolf) [inline src/game_objects.rs]: This is a weird method name: it makes sense only for Buildings. Suggestion: remove_first() (or just discard it and call remove(0)).

**Comment by Lycowolf** [comment]
You have quite a lot of methods named from_*. Consider using From trait (https://doc.rust-lang.org/std/convert/trait.From.html) instead.

File: src/game_objects.rs [hunk 2]
```diff
@@ -166,13 +219,14 @@ impl NumberMap {
...
     pub fn report(&self) {
         println!("Game state:");
-        for (key, val) in self.changed.iter().sorted() {
-            println!(" {}: {}", key, val);
-        }
+        //FIXME or delete
+        // for (key, val) in self.changed.iter().sorted() {
+        //     println!(" {}: {}", key, val);
+        // }
     }
```
Comment (Lycowolf) [inline src/game_objects.rs]: This method is useless: remove it

File: src/ui/mod.rs [hunk 3]
```diff
@@ -90,48 +86,50 @@ impl TakeTurnState {
...
         widgets.push(Box::new(zone));
 
         let base_numbers_position = Vector::new(4.0 * UI_UNIT, PLAYER_BOARD_FROM_TOP + 15.0 * UI_UNIT);
-        for (num, (name, value)) in gameplay_state.get_board().globals.iter().enumerate() {
+
+        for (num, currency) in Globals::in_game().iter().enumerate() {
+            let value = gameplay_state.get_board().globals.get(*currency);
             widgets.push(Box::new(Button::new(
-                format!("{}\n {}", name, value),
+                format!("{:?}\n {}", currency, value),
```
Comment (Lycowolf) [inline src/ui/mod.rs:121]: This uses Debug formatting for user-facing interface. This is wrong: you should implement Display. 
It is probably OK to implement it just as 

```
fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    // We know that debug formatting is OK
    <self as Debug>::fmt(f)
}
```

Comment (Lycowolf) [reply src/ui/mod.rs:121]: BTW: here you can see the rare left-swimming turbofish!

File: src/game_objects.rs [hunk 4]
```diff
@@ -44,80 +87,85 @@ pub struct Card {
...
+    }
+
+    pub fn all_effects(&self, event_selector: fn(&Card)-> &Vec<Effect>) -> Vec<(BoardZone, Card, Effect)> {
+
+        self.cards.iter()
+                .flat_map(|c| izip!(iter::repeat(self.zone), 
+                                    iter::repeat(c).cloned(),
+                                    event_selector(c).iter().cloned()) )
+                .collect()
+    }
```
Comment (Lycowolf) [inline src/game_objects.rs]: This requires a documentation comment (after looking at it, I was quite confused what it actually does).

Suggestion:

```rust
/// Extract a Vec of effects for each card in the container (effects can repeat). 
/// Join these lists together, tagging every effect with the card that causes it and zone the card belongs to.
///
/// Expected use: to evaluate on_turn_start effect etc.
```

File: src/game_objects.rs (hunk 1 above)
Comment (keltoff) [reply src/game_objects.rs]: Discarded, it is simple and used in one place only

File: src/game_objects.rs (hunk 2 above)
Comment (keltoff) [reply src/game_objects.rs]: done

File: src/ui/mod.rs (hunk 3 above)
Comment (keltoff) [reply src/ui/mod.rs:121]: ok, I'll use this

**Comment by keltoff** [comment]
Updated according to comments. I pretty much accepted all your suggestions.

**Pull request merged**